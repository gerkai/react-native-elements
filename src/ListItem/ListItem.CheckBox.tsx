import React from 'react';
import { StyleSheet } from 'react-native';
import CheckBox, { CheckBoxProps } from '../CheckBox';
import { RneFunctionComponent } from '../helpers';

export type ListItemCheckBoxProps = CheckBoxProps;

/** This allows adding CheckBox to the ListItem.
 * This, Receives all [CheckBox](checkbox.md#props) props. */
export const ListItemCheckBox: RneFunctionComponent<ListItemCheckBoxProps> = ({
  containerStyle,
  ...props
}) => {
  return (
    <CheckBox
      {...props}
      containerStyle={StyleSheet.flatten([styles.container, containerStyle])}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    margin: 0,
    marginRight: 0,
    marginLeft: 0,
    padding: 0,
  },
});

ListItemCheckBox.displayName = 'ListItem.CheckBox';
