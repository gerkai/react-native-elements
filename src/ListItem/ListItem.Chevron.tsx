import React from 'react';
import { StyleSheet, Platform } from 'react-native';
import { RneFunctionComponent } from '../helpers';
import Icon, { IconProps } from '../Icon';

export type ListItemChevronProps = Partial<IconProps>;

/** This allows adding a Chevron Icon(arrow) to the ListItem.
 * This, Receives all [Icon](icon.md#props) props. */
export const ListItemChevron: RneFunctionComponent<ListItemChevronProps> = ({
  containerStyle,
  ...props
}: Partial<IconProps>) => {
  return (
    <Icon
      type={Platform.OS === 'ios' ? 'ionicon' : 'material'}
      color="#D1D1D6"
      name={
        Platform.OS === 'ios'
          ? 'chevron-forward-outline'
          : 'keyboard-arrow-right'
      }
      size={16}
      containerStyle={StyleSheet.flatten([
        { alignSelf: 'center' },
        containerStyle,
      ])}
      {...props}
    />
  );
};

ListItemChevron.displayName = 'ListItem.Chevron';
