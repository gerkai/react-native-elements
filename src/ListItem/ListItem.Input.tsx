import React from 'react';
import { StyleSheet } from 'react-native';
import { RneFunctionComponent } from '../helpers';
import Input, { InputProps } from '../Input';

export type ListItemInputProps = InputProps;

/** This allows adding an Text Input within the ListItem.
 * This, Receives all [Input](input.md#props) props. */
export const ListItemInput: RneFunctionComponent<ListItemInputProps> = ({
  inputStyle,
  inputContainerStyle,
  containerStyle,
  ...props
}) => {
  return (
    <Input
      renderErrorMessage={false}
      {...props}
      inputStyle={StyleSheet.flatten([styles.input, inputStyle])}
      inputContainerStyle={StyleSheet.flatten([
        styles.inputContainer,
        inputContainerStyle,
      ])}
      containerStyle={StyleSheet.flatten([styles.container, containerStyle])}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingRight: 0,
  },
  inputContainer: {
    flex: 1,
    borderBottomWidth: 0,
    width: null,
    height: null,
  },
  input: {
    flex: 1,
    textAlign: 'right',
    width: null,
    height: null,
  },
});

ListItemInput.displayName = 'ListItem.Input';
