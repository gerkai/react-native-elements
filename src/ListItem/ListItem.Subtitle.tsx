import React from 'react';
import { StyleSheet, Platform, TextProps } from 'react-native';
import { RneFunctionComponent } from '../helpers';
import Text from '../Text';

const ANDROID_SECONDARY = 'rgba(0, 0, 0, 0.54)';

export type ListItemSubtitleProps = TextProps & { right?: boolean };

/** This allows adding SubTitle to the ListItem.
 * This, Receives all [Text](text.md#props) props. */
export const ListItemSubtitle: RneFunctionComponent<ListItemSubtitleProps> = ({
  style,
  right,
  children,
  ...props
}) => {
  return (
    <Text
      testID="listItemTitle"
      style={StyleSheet.flatten([
        styles.subtitle,
        right && styles.rightSubtitle,
        style,
      ])}
      {...props}
    >
      {children}
    </Text>
  );
};

const styles = StyleSheet.create({
  subtitle: {
    backgroundColor: 'transparent',
    ...Platform.select({
      ios: {
        fontSize: 15,
      },
      default: {
        color: ANDROID_SECONDARY,
        fontSize: 14,
      },
    }),
  },
  rightSubtitle: {
    color: ANDROID_SECONDARY,
  },
});

ListItemSubtitle.displayName = 'ListItem.Subtitle';
