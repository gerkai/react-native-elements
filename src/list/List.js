import PropTypes from 'prop-types';
import React from 'react';
import {
  View,
  StyleSheet,
  ViewPropTypes as RNViewPropTypes,
} from 'react-native';
import colors from '../config/colors';

const ViewPropTypes = RNViewPropTypes || View.propTypes;

const List = props => {
  const { children, containerStyle, ...attributes } = props;
  return (
    <View
      style={[styles.listContainer, containerStyle && containerStyle]}
      {...attributes}
    >
      {children}
    </View>
  );
};

List.propTypes = {
  children: PropTypes.any,
  containerStyle: ViewPropTypes.style,
};

const styles = StyleSheet.create({
  listContainer: {
    marginTop: 20,
    borderTopWidth: 0.5,
    borderColor: colors.greyOutline,
    backgroundColor: colors.white,
  },
});

export default List;
