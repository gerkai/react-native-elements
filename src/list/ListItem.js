import PropTypes from 'prop-types';
import React from 'react';
import {
  Platform,
  StyleSheet,
  Switch,
  TouchableHighlight,
  View,
} from 'react-native';
import renderNode from '../helpers/renderNode';
import nodeType from '../helpers/nodeType';
import Avatar from '../avatar/Avatar';
import Badge from '../badge/badge';
import CheckBox from '../checkbox/CheckBox';
import Icon from '../icons/Icon';
import Text from '../text/Text';
import ButtonGroup from '../buttons/ButtonGroup';
import Input from '../input/Input';
import ViewPropTypes from '../config/ViewPropTypes';
import colors from '../config/colors';

const ANDROID_SECONDARY = 'rgba(0, 0, 0, 0.54)';

const chevronDefaultProps = {
  type: Platform.OS === 'ios' ? 'ionicon' : 'material',
  color: '#D1D1D6',
  name: Platform.OS === 'ios' ? 'ios-arrow-forward' : 'keyboard-arrow-right',
  size: 16,
};
const checkmarkDefaultProps = {
  name: 'check',
  size: 20,
  color: colors.primary,
};
const renderText = (content, defaultProps, style) =>
  renderNode(Text, content, {
    ...defaultProps,
    style: [style, defaultProps && defaultProps.style],
  });
const renderAvatar = content =>
  renderNode(Avatar, content, {
    size: 40,
    rounded: true,
  });
const renderIcon = content =>
  renderNode(Icon, content, {
    color: Platform.OS === 'ios' ? null : ANDROID_SECONDARY,
    size: 24,
  });

const ListItem = props => {
  const {
    title,
    titleStyle,
    titleProps,
    subtitle,
    subtitleStyle,
    subtitleProps,
    containerStyle,
    onPress,
    onLongPress,
    component: Component = onPress || onLongPress ? TouchableHighlight : View,
    leftIcon,
    leftAvatar,
    leftElement,
    rightIcon,
    rightAvatar,
    rightElement,
    rightTitle,
    rightTitleStyle,
    rightTitleProps,
    rightSubtitle,
    rightSubtitleStyle,
    rightSubtitleProps,
    input,
    buttonGroup,
    switch: switchProps, // switch is a reserved keyword
    checkBox,
    badge,
    chevron,
    contentContainerStyle,
    rightContentContainerStyle,
    checkmark,
    disabled,
    disabledStyle,
    bottomDivider,
    topDivider,
    pad,
    linearGradientProps,
    ViewComponent = linearGradientProps && global.Expo
      ? global.Expo.LinearGradient
      : View,
    ...attributes
  } = props;
  return (
    <Component
      {...attributes}
      onPress={onPress}
      onLongPress={onLongPress}
      disabled={disabled}
    >
      <PadView
        Component={ViewComponent}
        {...linearGradientProps}
        style={[
          styles.container,
          (buttonGroup || switchProps) && { paddingVertical: 8 },
          topDivider && { borderTopWidth: StyleSheet.hairlineWidth },
          bottomDivider && { borderBottomWidth: StyleSheet.hairlineWidth },
          containerStyle,
          disabled && disabledStyle,
        ]}
        pad={pad}
      >
        {renderNode(Text, leftElement)}
        {renderIcon(leftIcon)}
        {renderAvatar(leftAvatar)}
        {(title || subtitle) && (
          <View style={[styles.contentContainer, contentContainerStyle]}>
            {renderText(title, titleProps, [styles.title, titleStyle])}
            {renderText(subtitle, subtitleProps, [
              styles.subtitle,
              subtitleStyle,
            ])}
          </View>
        )}
        {(!!rightTitle || !!rightSubtitle) && (
          <View
            style={[styles.rightContentContainer, rightContentContainerStyle]}
          >
            {renderText(rightTitle, rightTitleProps, [
              styles.title,
              styles.rightTitle,
              rightTitleStyle,
            ])}
            {renderText(rightSubtitle, rightSubtitleProps, [
              styles.subtitle,
              styles.rightSubtitle,
              rightSubtitleStyle,
            ])}
          </View>
        )}
        {input && (
          <Input
            {...input}
            inputStyle={[styles.input, input && input.inputStyle]}
            inputContainerStyle={[
              styles.inputContentContainer,
              input && input.inputContainerStyle,
            ]}
            containerStyle={[
              styles.inputContainer,
              input && input.containerStyle,
            ]}
          />
        )}
        {switchProps && <Switch {...switchProps} />}
        {checkBox && (
          <CheckBox
            {...checkBox}
            containerStyle={[
              styles.checkboxContainer,
              checkBox && checkBox.containerStyle,
            ]}
          />
        )}
        {badge && <Badge {...badge} />}
        {buttonGroup && (
          <ButtonGroup
            {...buttonGroup}
            containerStyle={[
              styles.buttonGroupContainer,
              buttonGroup && buttonGroup.containerStyle,
            ]}
          />
        )}
        {renderAvatar(rightAvatar)}
        {renderIcon(rightIcon)}
        {renderNode(Text, rightElement)}
        {renderNode(Icon, checkmark, checkmarkDefaultProps)}
        {renderNode(Icon, chevron, chevronDefaultProps)}
      </PadView>
    </Component>
  );
};

const Checkmark = ({ color }) => (
  <Icon type="material" name="check" size={20} color={color} />
);

const styles = StyleSheet.create({
  container: {
    ...Platform.select({
      ios: {
        padding: 14,
      },
      default: {
        padding: 16,
      },
    }),
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: 'white',
    borderColor: colors.divider,
  },
  title: {
    backgroundColor: 'transparent',
    ...Platform.select({
      ios: {
        fontSize: 17,
      },
      default: {
        fontSize: 16,
      },
    }),
  },
  subtitle: {
    backgroundColor: 'transparent',
    ...Platform.select({
      ios: {
        fontSize: 15,
      },
      default: {
        color: ANDROID_SECONDARY,
        fontSize: 14,
      },
    }),
  },
  contentContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  rightContentContainer: {
    flex: 0.5,
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  inputContainer: {
    flex: 1,
  },
  inputContentContainer: {
    flex: 1,
    borderBottomWidth: 0,
    width: null,
    height: null,
  },
  input: {
    flex: 1,
    textAlign: 'right',
    width: null,
    height: null,
    marginLeft: 0,
  },
  checkboxContainer: {
    margin: 0,
    marginRight: 0,
    marginLeft: 0,
    padding: 0,
  },
  buttonGroupContainer: {
    flex: 1,
    marginLeft: 0,
    marginRight: 0,
    marginTop: 0,
    marginBottom: 0,
  },
  rightTitle: {
    color: ANDROID_SECONDARY,
  },
  rightSubtitle: {
    color: ANDROID_SECONDARY,
  },
});

ListItem.propTypes = {
  containerStyle: ViewPropTypes.style,
  contentContainerStyle: ViewPropTypes.style,
  rightContentContainerStyle: ViewPropTypes.style,
  component: PropTypes.func,
  onPress: PropTypes.func,
  onLongPress: PropTypes.func,
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  titleStyle: Text.propTypes.style,
  titleProps: PropTypes.object,
  subtitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  subtitleStyle: Text.propTypes.style,
  subtitleProps: PropTypes.object,
  leftIcon: nodeType,
  leftAvatar: nodeType,
  leftElement: nodeType,
  rightIcon: nodeType,
  rightAvatar: nodeType,
  rightElement: nodeType,
  rightTitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  rightTitleStyle: Text.propTypes.style,
  rightTitleProps: PropTypes.object,
  rightSubtitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  rightSubtitleStyle: Text.propTypes.style,
  rightSubtitleProps: PropTypes.object,
  input: PropTypes.object,
  buttonGroup: PropTypes.object,
  switch: PropTypes.object,
  checkBox: PropTypes.object,
  badge: PropTypes.object,
  chevron: nodeType,
  checkmark: nodeType,
  disabled: PropTypes.bool,
  disabledStyle: ViewPropTypes.style,
  topDivider: PropTypes.bool,
  bottomDivider: PropTypes.bool,
  pad: PropTypes.number,
  linearGradientProps: PropTypes.object,
  ViewComponent: PropTypes.func,
};

ListItem.defaultProps = {
  pad: 16,
};

const PadView = ({ children, pad, Component, ...props }) => {
  const childrens = React.Children.toArray(children);
  const length = childrens.length;
  const Container = Component || View;
  return (
    <Container {...props}>
      {React.Children.map(
        childrens,
        (child, index) =>
          child && [child, index !== length - 1 && <View width={pad} />]
      )}
    </Container>
  );
};

export default ListItem;
