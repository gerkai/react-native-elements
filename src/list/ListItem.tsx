import { withTheme } from '../config';
import { ListItemProps, ListItemBase } from './ListItemBase';
import ListItemContent from './ListItemContent';
import ListItemChevron from './ListItemChevron';
import ListItemInput from './ListItemInput';
import ListItemCheckBox from './ListItemCheckBox';
import ListItemButtonGroup from './ListItemButtonGroup';
import ListItemTitle from './ListItemTitle';
import ListItemSubtitle from './ListItemSubtitle';
import ListItemSwipeable from './ListItemSwipeable';
import ListItemAccordion from './ListItemAccordion';
import { RneFunctionComponent } from '../helpers';

interface ListItem extends RneFunctionComponent<ListItemProps> {
  Accordion: typeof ListItemAccordion;
  Chevron: typeof ListItemChevron;
  Content: typeof ListItemContent;
  Input: typeof ListItemInput;
  Title: typeof ListItemTitle;
  Subtitle: typeof ListItemSubtitle;
  Swipeable: typeof ListItemSwipeable;
  CheckBox: typeof ListItemCheckBox;
  ButtonGroup: typeof ListItemButtonGroup;
}

const ListItem: ListItem = Object.assign(ListItemBase);

export { ListItem };

const ThemedListItem = Object.assign(withTheme(ListItem, 'ListItem'), {
  Accordion: ListItemAccordion,
  Chevron: ListItemChevron,
  Content: ListItemContent,
  Input: ListItemInput,
  Title: ListItemTitle,
  Subtitle: ListItemSubtitle,
  Swipeable: ListItemSwipeable,
  CheckBox: ListItemCheckBox,
  ButtonGroup: ListItemButtonGroup,
});

export default ThemedListItem;
