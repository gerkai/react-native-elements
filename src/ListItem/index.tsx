import { withTheme } from '../config';
import { ListItemBaseProps, ListItemBase } from './ListItem';
import { ListItemContent } from './ListItem.Content';
import { ListItemChevron } from './ListItem.Chevron';
import { ListItemInput } from './ListItem.Input';
import { ListItemCheckBox } from './ListItem.CheckBox';
import { ListItemButtonGroup } from './ListItem.ButtonGroup';
import { ListItemTitle } from './ListItem.Title';
import { ListItemSubtitle } from './ListItem.Subtitle';
import {
  ListItemSwipeable,
  ListItemSwipeableProps,
} from './ListItem.Swipeable';
import {
  ListItemAccordion,
  ListItemAccordionProps,
} from './ListItem.Accordion';
import { RneFunctionComponent } from '../helpers';

const ThemedListItemContent = withTheme(ListItemContent, 'ListItemContent');
const ThemedListItemChevron = withTheme(ListItemChevron, 'ListItemChevron');
const ThemedListItemInput = withTheme(ListItemInput, 'ListItemInput');
const ThemedListItemCheckBox = withTheme(ListItemCheckBox, 'ListItemCheckBox');
const ThemedListItemButtonGroup = withTheme(
  ListItemButtonGroup,
  'ListItemButtonGroup'
);
const ThemedListItemTitle = withTheme(ListItemTitle, 'ListItemTitle');
const ThemedListItemSubtitle = withTheme(ListItemSubtitle, 'ListItemSubtitle');
const ThemedListItemSwipeable = withTheme(
  ListItemSwipeable,
  'ListItemSwipeable'
);
const ThemedListItemAccordion = withTheme(
  ListItemAccordion,
  'ListItemAccordion'
);

export type ListItemProps = RneFunctionComponent<ListItemBaseProps> & {
  Accordion: typeof ThemedListItemAccordion;
  Chevron: typeof ThemedListItemChevron;
  Content: typeof ThemedListItemContent;
  Input: typeof ThemedListItemInput;
  Title: typeof ThemedListItemTitle;
  Subtitle: typeof ThemedListItemSubtitle;
  Swipeable: typeof ThemedListItemSwipeable;
  CheckBox: typeof ThemedListItemCheckBox;
  ButtonGroup: typeof ThemedListItemButtonGroup;
};

export const ListItem: ListItemProps = Object.assign(ListItemBase);

export type { ListItemAccordionProps, ListItemSwipeableProps };

const ThemedListItem = Object.assign(withTheme(ListItem, 'ListItem'), {
  Accordion: ThemedListItemAccordion,
  Chevron: ThemedListItemChevron,
  Content: ThemedListItemContent,
  Input: ThemedListItemInput,
  Title: ThemedListItemTitle,
  Subtitle: ThemedListItemSubtitle,
  Swipeable: ThemedListItemSwipeable,
  CheckBox: ThemedListItemCheckBox,
  ButtonGroup: ThemedListItemButtonGroup,
});

export default ThemedListItem;
