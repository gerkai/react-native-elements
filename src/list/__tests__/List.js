import React from 'react';
import {shallow} from 'enzyme';
import List from '../List';
import ListItem from '../ListItem';

describe('List component', () => {
  it('should render without issues', () => {
    const component = shallow(<List />);

    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('should render children', () => {
    const component = shallow(<List><ListItem /></List>);

    expect(component.find('ListItem').length).toBe(1);
  });
});
