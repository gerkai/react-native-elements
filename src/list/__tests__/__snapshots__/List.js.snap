// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render without issues 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderBottomWidth": 1,
        "borderColor": "#cbd2d9",
        "borderTopWidth": 1,
        "marginTop": 20,
      },
      undefined,
    ]
  }
/>
`;
